### PR TITLE
Logical Memory Spaces

### PR DESCRIPTION
**Pull request by DavidPoliakoff**
Goal: allow for memory spaces which are functionally identical to some other spaces, but which are differentiable by the Profiling system, and by template specializations.

This is an excavation of an old approach to the problem I closed out a while ago. The approach is to use a Memory Space template, templated on

1) A name (likely to be removed)
2) The base memory space, the one on which memory operations functionally happen
3) A DefaultExecutionSpace (by default, the default execution space of the underlying memory space)
4) Whether the logical space inherits the access characteristics of the base space (default true). If false, the user can override the MemorySpaceAccess templates to specify where this memory spaces should be accessible from

In an initial state, I'm mostly putting this up just to document what it is, when I remove it from draft it should be ready for code review, concept review is welcome at any time

### CODE REVIEW

**Comment by DavidPoliakoff** [comment]
Retest this please

**Comment by DavidPoliakoff** [comment]
Retest this please

File: core/src/Kokkos_LogicalMemorySpace.hpp [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+
+  using execution_space =
+      typename std::conditional<std::is_void<DefaultExecutionSpace>::value,
+                                typename BaseSpace::execution_space,
+                                DefaultExecutionSpace>::type;
+
+  typedef Kokkos::Device<execution_space, memory_space> device_type;
+
+  /**\brief  Default memory space instance */
+  LogicalMemorySpace() : underlying_allocator(){};
```
Comment (dalg24) [inline core/src/Kokkos_LogicalMemorySpace.hpp:99]: Why is that one not defaulted?

**Comment by DavidPoliakoff** [comment]
> No unit test?!

Oddly, I need a tool test to properly test this. I might run some ideas by you soon

File: core/src/Kokkos_LogicalMemorySpace.hpp [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+
+template <class Namer, class BaseSpace, class DefaultExecutionSpace = void,
+          bool SharesAccessWithBase = true>
+class LogicalMemorySpace {
+ public:
+  //! Tag this class as a kokkos memory space
+  typedef LogicalMemorySpace<Namer, BaseSpace, DefaultExecutionSpace,
+                             SharesAccessWithBase>
+      memory_space;
+  typedef size_t size_type;
```
[the remark this answers is not in the document]
Comment (DavidPoliakoff) [reply core/src/Kokkos_LogicalMemorySpace.hpp:82]: I was just doing this out of respect for Daisy's aesthetic sensibilities, but sure :)

File: core/src/Kokkos_LogicalMemorySpace.hpp (hunk 1 above)
Comment (DavidPoliakoff) [reply core/src/Kokkos_LogicalMemorySpace.hpp:99]: Because I didn't think to, good comment, I'll do that

**Comment by DavidPoliakoff** [comment]
This implementation is actually broken and insufficient, I'll add a new one soon (with tests this time!)

**Pull request closed without merging**